### ISSUE TITLE
Can't scale deployment more than 1 replica

### ISSUE THREAD

**Issue by LukeChannings** [haproxytech/helm-charts#15]
If I set `replicaCount` to `1`, I have a successful deployment and everything is functional.
However, if I set it to 2 or more, one pod will start up, and the others will eventually CrashLoopBackoff because they fail their healthcheck.

Upon further investigation (shell on the pod, logs, shell on an ephemeral pod to confirm, etc.), it turns out that the other pods simply don't start haproxy. It's not listening on any ports, and an inspection of `top` shows it's not running.

Logs from the bad pod do not indicate any error:

```
2020/05/05 08:46:43
 _   _    _    ____
| | | |  / \  |  _ \ _ __ _____  ___   _
| |_| | / _ \ | |_) | '__/ _ \ \/ / | | |
|  _  |/ ___ \|  __/| | | (_) >  <| |_| |
|_| |_/_/   \_\_|   |_|  \___/_/\_\\__, |
 _  __     _                       |___/             ___ ____
| |/ /   _| |__   ___ _ __ _ __   ___| |_ ___  ___  |_ _/ ___|
| ' / | | | '_ \ / _ \ '__| '_ \ / _ \ __/ _ \/ __|  | | |
| . \ |_| | |_) |  __/ |  | | | |  __/ ||  __/\__ \  | | |___
|_|\_\__,_|_.__/ \___|_|  |_| |_|\___|\__\___||___/ |___\____|


2020/05/05 08:46:43 HAProxy Ingress Controller v1.4.2 46127a7

2020/05/05 08:46:43 Build from: git@github.com:haproxytech/kubernetes-ingress.git
2020/05/05 08:46:43 Build date: 2020-04-03T20:10:18

2020/05/05 08:46:43 ConfigMap: haproxy-ingress/haproxy-ingress-kubernetes-ingress
2020/05/05 08:46:43 Ingress class: internal
2020/05/05 08:46:43 Publish service: haproxy-ingress/haproxy-ingress-kubernetes-ingress
2020/05/05 08:46:43 Default backend service: haproxy-ingress/haproxy-ingress-kubernetes-ingress-default-backend
2020/05/05 08:46:43 Default ssl certificate: haproxy-ingress/haproxy-default-tls
2020/05/05 08:46:43 Running with  HA-Proxy version 2.0.14 2020/04/02 - https://haproxy.org/
2020/05/05 08:46:43 Starting HAProxy with /etc/haproxy/haproxy.cfg
2020/05/05 08:46:43 Running on haproxy-ingress-kubernetes-ingress-659b898986-hjpk9
[WARNING] 125/084643 (24) : Can't open server state file '/var/state/haproxy/global': No such file or directory
2020/05/05 08:46:43 Running on Kubernetes version: v1.17.4+k3s1 linux/amd64
```

Compared to the healthy pod:

```
2020/05/05 06:59:20
 _   _    _    ____
| | | |  / \  |  _ \ _ __ _____  ___   _
| |_| | / _ \ | |_) | '__/ _ \ \/ / | | |
|  _  |/ ___ \|  __/| | | (_) >  <| |_| |
|_| |_/_/   \_\_|   |_|  \___/_/\_\\__, |
 _  __     _                       |___/             ___ ____
| |/ /   _| |__   ___ _ __ _ __   ___| |_ ___  ___  |_ _/ ___|
| ' / | | | '_ \ / _ \ '__| '_ \ / _ \ __/ _ \/ __|  | | |
| . \ |_| | |_) |  __/ |  | | | |  __/ ||  __/\__ \  | | |___
|_|\_\__,_|_.__/ \___|_|  |_| |_|\___|\__\___||___/ |___\____|


2020/05/05 06:59:20 HAProxy Ingress Controller v1.4.2 46127a7

2020/05/05 06:59:20 Build from: git@github.com:haproxytech/kubernetes-ingress.git
2020/05/05 06:59:20 Build date: 2020-04-03T20:10:18

2020/05/05 06:59:20 ConfigMap: haproxy-ingress/haproxy-ingress-kubernetes-ingress
2020/05/05 06:59:20 Ingress class: internal
2020/05/05 06:59:20 Publish service: haproxy-ingress/haproxy-ingress-kubernetes-ingress
2020/05/05 06:59:20 Default backend service: haproxy-ingress/haproxy-ingress-kubernetes-ingress-default-backend
2020/05/05 06:59:20 Default ssl certificate: haproxy-ingress/haproxy-default-tls
2020/05/05 06:59:20 Running with  HA-Proxy version 2.0.14 2020/04/02 - https://haproxy.org/
2020/05/05 06:59:20 Starting HAProxy with /etc/haproxy/haproxy.cfg
2020/05/05 06:59:20 Running on haproxy-ingress-kubernetes-ingress-659b898986-42qmr
[WARNING] 125/065920 (23) : Can't open server state file '/var/state/haproxy/global': No such file or directory
2020/05/05 06:59:20 Running on Kubernetes version: v1.17.4+k3s1 linux/amd64
[NOTICE] 125/065920 (35) : New worker #1 (36) forked
2020/05/05 06:59:27 Confiugring default_backend haproxy-ingress-kubernetes-ingress-default-backend from ingress DefaultService
2020/05/05 06:59:27 successful update of LoadBalancer status of ingress home-assistant/home-assistant
2020/05/05 06:59:27 successful update of LoadBalancer status of ingress ***/***
2020/05/05 06:59:27 successful update of LoadBalancer status of ingress ***/***
2020/05/05 06:59:27 successful update of LoadBalancer status of ingress ***/***
2020/05/05 06:59:27 successful update of LoadBalancer status of ingress ***/***
2020/05/05 06:59:27 successful update of LoadBalancer status of ingress ***/***
2020/05/05 06:59:28 successful update of LoadBalancer status of ingress ***/***
2020/05/05 06:59:28 successful update of LoadBalancer status of ingress ***/***
2020/05/05 06:59:28 successful update of LoadBalancer status of ingress ***/***
2020/05/05 06:59:29 successful update of LoadBalancer status of ingress ***/***
2020/05/05 06:59:29 HAProxy reloaded
[NOTICE] 125/065929 (48) : New worker #1 (49) forked
2020/05/05 08:46:04 Modified: ***/*** - SRV_rGwUZ - ready
2020/05/05 08:46:04 Modified: ***/*** - SRV_AAnIs - maint
2020/05/05 08:46:04 Modified: ***/*** - SRV_P8INE - ready
2020/05/05 08:46:04 Modified: ***/***- SRV_klqWx - maint
2020/05/05 09:00:45 Modified: ***/*** - SRV_rGwUZ - maint
2020/05/05 09:00:45 Modified: ***/*** - SRV_P8INE - maint
2020/05/05 09:00:54 Modified: ***/*** - SRV_zYmkr - ready
2020/05/05 09:00:54 Modified: ***/*** - SRV_38MlW - ready
```

Here's my full helm values:

```yaml
---
kubernetes-ingress:
  controller:
    kind: Deployment
    replicaCount: 1
    ingressClass: internal
    defaultTLSSecret:
      enabled: true
      secret: haproxy-ingress/haproxy-default-tls
    service:
      type: NodePort
      labels:
        app: haproxy-ingress
    publishService:
      enabled: true
    nodeSelector:
      kubernetes.io/arch: amd64
  defaultBackend:
    replicaCount: 1
    nodeSelector:
      kubernetes.io/arch: amd64
```

**Comment by Mo3m3n** [comment]
That's quite strange.
Can you try `haproxytech/kubernetes-ingress:dev` image this is a build of what is currently in master. In case this is a bug that was previously fixed. It seems to not be the case though
With the dev image you can also use the CLI arg `--log debug` to get more logs from the controller.
By the way this logging enhancement will be available in next release and will make it easier to setup via Helm.
If you want to share the logs in a more private way you can join our [Community Slack](https://github.com/haproxytech/kubernetes-ingress#discussion)

**Comment by LukeChannings** [comment]
Hey, I tried to join the Slack, but it said I need an @haproxy.com email address. I might be doing something wrong!

Switched to the dev image and no change. Logs:

```
2020/05/05 14:46:08
 _   _    _    ____
| | | |  / \  |  _ \ _ __ _____  ___   _
| |_| | / _ \ | |_) | '__/ _ \ \/ / | | |
|  _  |/ ___ \|  __/| | | (_) >  <| |_| |
|_| |_/_/   \_\_|   |_|  \___/_/\_\\__, |
 _  __     _                       |___/             ___ ____
| |/ /   _| |__   ___ _ __ _ __   ___| |_ ___  ___  |_ _/ ___|
| ' / | | | '_ \ / _ \ '__| '_ \ / _ \ __/ _ \/ __|  | | |
| . \ |_| | |_) |  __/ |  | | | |  __/ ||  __/\__ \  | | |___
|_|\_\__,_|_.__/ \___|_|  |_| |_|\___|\__\___||___/ |___\____|

2020/05/05 14:46:08 HAProxy Ingress Controller v1.4.3 7739fa8.dev
2020/05/05 14:46:08 Build from: https://github.com/haproxytech/kubernetes-ingress
2020/05/05 14:46:08 Build date: 2020-05-05T13:22:39
2020/05/05 14:46:08 ConfigMap: haproxy-ingress/haproxy-ingress-kubernetes-ingress
2020/05/05 14:46:08 Ingress class: internal
2020/05/05 14:46:08 Publish service: haproxy-ingress/haproxy-ingress-kubernetes-ingress
2020/05/05 14:46:08 Default backend service: haproxy-ingress/haproxy-ingress-kubernetes-ingress-default-backend
2020/05/05 14:46:08 Default ssl certificate: haproxy-ingress/haproxy-default-tls
2020/05/05 14:46:08 Controller sync period: 5s
2020/05/05 14:46:09 INFO    controller.go:264 Running with HA-Proxy version 2.0.14 2020/04/02 - https://haproxy.org/
2020/05/05 14:46:09 INFO    controller.go:269 Starting HAProxy with /etc/haproxy/haproxy.cfg
2020/05/05 14:46:09 INFO    controller.go:274 Running on haproxy-ingress-kubernetes-ingress-k755c
[WARNING] 125/144609 (23) : Can't open server state file '/var/state/haproxy/global': No such file or directory
2020/05/05 14:46:09 INFO    controller.go:108 Running on Kubernetes version: v1.17.4+k3s1 linux/amd64
2020/05/05 14:46:16 ERROR   controller.go:332 /var/run/haproxy-runtime-api.sock dial unix /var/run/haproxy-runtime-api.sock: connect: no such
 file or directory
2020/05/05 14:46:16 ERROR   controller.go:334 haproxy is not running, trying to start it
2020/05/05 14:46:16 INFO    controller.go:226 HAProxy reloaded
[WARNING] 125/144616 (28) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[WARNING] 125/144616 (28) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[WARNING] 125/144616 (28) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[WARNING] 125/144616 (28) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[WARNING] 125/144616 (28) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[WARNING] 125/144616 (28) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[WARNING] 125/144616 (28) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[WARNING] 125/144616 (28) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[WARNING] 125/144616 (28) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[WARNING] 125/144616 (28) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[WARNING] 125/144616 (28) : Can't open server state file '/var/state/haproxy/global': No such file or directory
[WARNING] 125/144616 (28) : Can't open server state file '/var/state/haproxy/global': No such file or directory
```

Confirmed the image it's running is:

`'docker.io/haproxytech/kubernetes-ingress:dev'`
`docker.io/haproxytech/kubernetes-ingress@sha256:82bf4771e5072c7ab43652e2a12dfb6c570b88c992c6700225b4746bf90e85aa`

**Comment by LukeChannings** [comment]
I switched to a DaemonSet for this, no change. 

Here's the full pod manifest:

[pod-manifest.yaml.txt](https://github.com/haproxytech/helm-charts/files/4581461/pod-manifest.yaml.txt)

**Comment by Mo3m3n** [comment]
>ingress@sha256:82bf4771e5072c7ab43652e2a12dfb6c570b88c992c6700225b4746bf90e85aa

Correct. Sorry should have sent that in the first place.

>Hey, I tried to join the Slack, but it said I need an @haproxy.com email address. I might be doing something wrong!

The community slack haproxy.slack.com should not ask for @haproxy.com email  (to not confuse with corportae slack haproxytech.slack.com)

>Logs:

Interesting It seems like HAProxy is not able to run. If one pod is able to do so and not the other this is typical of a port conflict that happens if pods are running on the same node and HostPort/HostNetwork is enabled.
Now it is time to get more info about HAproxy itself by enabling HAProxy logs this time (in addition to controller logs), you can find documentation about this in the controller [repo](https://github.com/haproxytech/kubernetes-ingress/tree/master/documentation#logging).
Use **stdout** logs set this quickly.

**Comment by LukeChannings** [comment]
Hmm, I can't find a way to do it without an `@haproxy.com` address from haproxy.slack.com.

<img width="668" alt="Screenshot 2020-05-05 at 16 28 45" src="https://user-images.githubusercontent.com/461449/81087657-e9438400-8ef1-11ea-97c8-bca92bcc339f.png">

I did a few things to debug:

- Enabled `--log trace` on the controller
- Set the liveliness probes to infinity so that I could `k exec` onto the failed node.

I have found:
- The haproxy.cfg between the healthy pod and the unhealthy pod is almost identical
- Running `haproxy -W -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -sf 28 -x /var/run/haproxy-runtime-api.sock` directly gives me `[WARNING] 125/155031 (48) : Can't open server state file '/var/state/haproxy/global': No such file or directory` messages
- Writing `1` into `/var/state/haproxy/global` gets rid of the errors, but then haproxy does not log anything at all.

I edited the haproxy.cfg and made sure there was:

```
global
    log 127.0.0.1 local0 debug

defaults
    log global
```

Still no dice, haproxy doesn't log anything. Not really sure what to try next.

**Comment by Mo3m3n** [comment]
>Hmm, I can't find a way to do it without an @haproxy.com address from haproxy.slack.com.

Sorry  wrong link there .. it should be slack.haproxy.org, I am going to fix that.

> Writing 1 into /var/state/haproxy/global gets rid of the errors, but then haproxy does not log anything at all.

Yes that's only happen once when `var/state/haproxy/global` you can ignore it eventhough it is waiting for a proper fix ^^

>I edited the haproxy.cfg and made sure there was:

You don't have to edit haproxy.cfg by hand but rather use the [syslog-server](https://github.com/haproxytech/kubernetes-ingress/tree/master/documentation#logging) annotation by hand.

**Comment by LukeChannings** [comment]
After extensive debugging with the help of @Mo3m3n we found that this issue was caused by a memory leak in `haproxy`, seemingly caused by my specific configuration.

I destroyed my cluster and re-installed the controller (and other workloads), and the issue has not recurred. I am running Kubernetes on k3s and commodity hardware, over two different Linux distributions, and with metallb

It's possible that the interaction of all of these factors lead to this bug, but we might never know. If it recurs and a more obvious cause presents itself I will re-open this issue.

Thanks very much for all of the help debugging this!
